### PR TITLE
Generate message code using md5()

### PR DESCRIPTION
**Pull request by mjauvin**
ref. https://github.com/rainlab/translate-plugin/pull/566

### CODE REVIEW

File: updates/version.yaml [hunk 1]
```diff
@@ -103,4 +103,3 @@
 "2.0.0":
     - Rebrand to Winter.Translate
     - v2.0.0/rename_tables.php
-
```
Comment (LukeTowers) [inline updates/version.yaml:106]: Migration is missing

Comment (mjauvin) [reply updates/version.yaml:106]: oops! fixed.

**Comment by LukeTowers** [comment]
@mjauvin that last commit should be on the main branch I think

File: updates/version.yaml [hunk 2]
```diff
@@ -104,4 +104,8 @@
     - Rebrand to Winter.Translate
     - Fix location for dropdown-to in css file
     - v2.0.0/rename_tables.php
-
+"2.1.0":
+    - |
```
Comment (LukeTowers) [inline updates/version.yaml:108]: What's with the bar character? Better to just have it be one line for version descriptions.

Comment (mjauvin) [reply updates/version.yaml:108]: It's more obvious when the migration runs with the output on two lines.

The second line has the "See https://github.com/wintercms/wn-translate-plugin/pull/2"

**Comment by mahony0** [comment]
@mjauvin I've used md5 keys for translations on my 2011 made custom local codebase so it looks very familiar to me. Besides md5 collisions, this is a very effective way to use (especially for fixed key size).

mb_strtolower can't always lowering non-unicode characters. Check this for turkish custom characters:

```php
$upper = trim('IİĞÜŞÖÇ');
$lower = trim('ıiğüşöç');

$upperMd5 = md5(mb_strtolower($upper, 'UTF-8'));
$lowerMd5 = md5(mb_strtolower($lower, 'UTF-8'));

var_dump($upper);
var_dump($lower);

var_dump($upperMd5);
var_dump($lowerMd5);
```

this gives something like:
```php
// string(13) "IİĞÜŞÖÇ"
// string(13) "ıiğüşöç"
// string(32) "34639259f8e609370a6364e74119c975"
// string(32) "83f0d9e5f7fabdcc68c0b3aa606e6507"
```

as this is not correct, I've solved this by converting custom characters beforehand:
```php
$upper = str_replace(['I', 'İ'], ['ı', 'i'], $upper);
```

As you can see, this fix is just for turkish characters and I don't know exactly which languages and which characters are wrongly converting in php.

**Comment by mjauvin** [comment]
I understand mb_strtolower() will not lead to the same key in all cases, but unless you have a different method, this is the best solution.

In those rare cases where it differs, multiple translations will be needed.

You can also use a key instead of actual text and provide translations for all languages on the "Translate messages" page to overcome any problems with this.

Thanks.

**Comment by mahony0** [comment]
Actually, why we need mb_strtolower anyway? If developer hardcoding string as capital letters why we are converting it to the lowercase?

I think 'HOMEPAGE', 'Homepage' or 'HomePage' are really not the same about their use cases. By lowercasing it, there's no possibility to translate these words as different texts.

Hope I shed some light on actual issue.

**Comment by mahony0** [comment]
Yes, I mean both of these should not be use the same message translation.

I'm proposing that these 'Hello' and 'hello' strings should have separate entries on backend "Translate messages" page if applicable.

**Comment by matteotrubini** [comment]
@mahony0 is right.

Simplifying: `{{ 'HELLO' | _ }}` and `{{ 'hello' | _ }}` should create different records to ensure 1:1 correspondency between originals and translations.

You can always use `{{ 'HELLO' | _ }}` and `{{ 'hello' | _ | uppercase }} ` to deduplicate translations.
NB: in this example I'm assuming to use [Twig extensions plugin
](https://github.com/vojtasvoboda/oc-twigextensions-plugin)

**Comment by mjauvin** [comment]
I agree but I was trying to stay as close to previous behavior as possible, which was using "strtolower()" on the key

**Comment by matteotrubini** [comment]
I think that mb_strtolower() was used to ensure code case normalization (Ex about MySQL: https://stackoverflow.com/questions/3936967/mysql-is-a-select-statement-case-sensitive), which is no more necessary using hashes.

I'm thinking about a BC: if anyone has used "HELLO" and "hello" at the same time: now are coded with the key, then will be different md5 so "HELLO" (non-lowercase version) will miss translation.

We can manage this on migration this creating both "regular" and lowercase md5 hashes (if dirrerent) to preserve both "translation".

Then docs should inform that translation may have been duplicated on upgrade, but from that point - and for new ones - a new case sensitive management is on stage.

**Comment by mjauvin** [comment]
If you provide me with the migration, I'll apply the change to remove the mb_strtolower()

**Comment by matteotrubini** [comment]
WIP

**Comment by matteotrubini** [comment]
Is there any plan to merge this PR soon (maybe togheter with `mb_strtolower()` removal and #6 ) ?
I'm facing a lot of troubles managing translation with the old `makeMessageCode()`

**Comment by LukeTowers** [comment]
@matteotrubini could you provide examples of the issues that you're running into with the current makeMessageCode()? I haven't seen @mjauvin around in a while so I'll have to get reacquainted with the issue and the proposed solution so that I can merge this.

**Comment by matteotrubini** [comment]
- An old thought rainlab/translate-plugin#256
- A more recent one rainlab/translate-plugin#563
- Currently if 2 strings (like sentences) is longer than 250 chars, but varies only from the 251° char, keys will collide.

About this last point, just today I have to manage something like this:

`{{ '<a data-toggle="tooltip" title="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse ac elit vulputate, dictum ligula at, tempus libero. Nam pretium in urna id dictum.">Lorem ipsum dolor sit amet</a>, consectetur adipiscing elit. Suspendisse ac elit vulputate, dictum ligula at, tempus libero. Nam pretium in urna id dictum. Phasellus malesuada felis nisi, quis iaculis nibh tincidunt non. Aenean vulputate vehicula ligula ut vulputate.'|_ }}`

`{{ '<a data-toggle="tooltip" title="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse ac elit vulputate, dictum ligula at, tempus libero. Nam pretium in urna id dictum.">Lorem ipsum dolor sit amet</a>, consectetur adipiscing elit. Suspendisse ac elit vulputate, dictum ligula at, tempus libero. Morbi iaculis mi arcu, eget commodo odio rutrum sit amet. Ut interdum, tortor eget congue iaculis, risus ligula tempus arcu, in venenatis erat ipsum eu erat.'|_ }}`

These snippets produce the same message code. And I cannot use translatable contents as them will complicate translation process (translatable strings are not consolidated in a single point exportable/importable).

Using an hash as key is definitely the safest thing to do. And assuming that md5 hashes will not collide in our use case, this will be enough (quicker and lighter on db rather than sha1).

> if anyone has used "HELLO" and "hello" at the same time: now are coded with the key, then will be different md5 so "HELLO" (non-lowercase version) will miss translation.
>
>We can manage this on migration this creating both "regular" and lowercase md5 hashes (if dirrerent) to preserve both "translation".

#6 will cover this case.

**Pull request merged**